File: 1002.find-common-characters.java
/*
 * @lc app=leetcode id=1002 lang=java
 *
 * [1002] Find Common Characters
 */

// @lc code=start

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<String> commonChars(String[] words) {
        List<String> ret = new ArrayList<>();
        if (words.length == 1) {
            for (char ch : words[0].toCharArray()) ret.add(String.valueOf(ch));
            return ret;
        }
        List<Map<Character, Integer>> maps = new ArrayList<>();
        for (String word : words) {
            maps.add(count(word));
        }
        for (char ch = 'a'; ch <= 'z'; ++ch) {
            final char _ch = ch;
            int times = (int) maps.stream()
                .map(m -> m.getOrDefault(_ch, 0))
                .mapToInt(x -> x)
                .min().orElse(-1);
            while (times != 0) {
                ret.add(String.valueOf(ch));
                times -= 1;
            }
        }
        return ret;
    }
    private Map<Character, Integer> count(String s) {
        Map<Character, Integer> ret = new HashMap<>();
        for (char ch : s.toCharArray()) {
            ret.put(ch, ret.getOrDefault(ch, 0) + 1);
        }
        return ret;
    }
}
// @lc code=end

File: 1002.find-common-characters.cpp
/*
 * @lc app=leetcode id=1002 lang=cpp
 *
 * [1002] Find Common Characters
 */

// @lc code=start

#include <algorithm>
#include <array>
#include <numeric>
#include <string>
#include <vector>
#include <limits>

using namespace std;

class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        using container_t = array<unsigned, 26>;
        vector<container_t> stats;
        transform(words.begin(), words.end(), back_inserter(stats), [&](const string& s) {
            container_t container = {0};
            for_each(s.begin(), s.end(), [&](char ch) { container[ch - 'a'] += 1; });
            return container;
        });
        container_t initial;
        initial.fill(numeric_limits<unsigned>::max());
        container_t res = accumulate(stats.begin(), stats.end(), initial, [](const container_t &acc, const container_t &curr) {
            container_t m = {0};
            transform(acc.begin(), acc.end(), curr.begin(), m.begin(), [](unsigned a, unsigned b) {
                return min(a, b);
            });
            return m;
        });
        vector<string> ret;
        for (auto ix = 0; ix != res.size(); ++ix) {
            while (res[ix] != 0) {
                ret.push_back(string(1, static_cast<char>('a' + ix)));
                res[ix] -= 1;
            }
        }
        return ret;
    }
};
// @lc code=end
